inst/include/dust/dust.hpp

```diff
@@ -316,12 +420,59 @@ class Dust {
   }
 
   void run(const size_t step_end) {
+    refresh_host();
 #ifdef _OPENMP
     #pragma omp parallel for schedule(static) num_threads(_n_threads)
 #endif
     for (size_t i = 0; i < _particles.size(); ++i) {
       _particles[i].run(step_end, _rng.state(i));
     }
+    _stale_device = true;
+  }
+
+  template <typename U = T>
+  typename std::enable_if<!dust::has_gpu_support<U>::value, void>::type
+  run_device(const size_t step_end) {
+    throw std::invalid_argument("GPU support not enabled for this object");
+  }
+
+  template <typename U = T>
+  typename std::enable_if<dust::has_gpu_support<U>::value, void>::type
+  run_device(const size_t step_end) {
+    refresh_device();
+    const size_t n_int = dust::device_work_size_int<T>(_shared[0]);
+    const size_t n_real = dust::device_work_size_real<T>(_shared[0]);
+    const size_t step_start = step();
+
+    // Allocation for our internal storage. We won't ever move this
+    // back to the cpu
+    const size_t n_particles = _particles.size();
+    const size_t n_state = n_state_full();
+    std::vector<real_t> y_next(n_particles * n_state);
+    std::vector<int> internal_int(n_particles * n_int);
+    std::vector<real_t> internal_real(n_particles * n_real);
+
+    run_particles<T>(step_start, step_end, n_particles,
+                     n_state, n_int, n_real, n_pars_effective(),
+                     _device_data.y.data(), y_next.data(),
+                     internal_int.data(), internal_real.data(),
+                     _shared, _device_data.rng.data());
+
+    // In the inner loop, the swap will keep the locally scoped
+    // interleaved variables updated, but the interleaved variables
+    // passed in have not yet been updated.  If an even number of
+    // steps have been run state will have been swapped back into the
+    // original place, but an on odd number of steps the passed
+    // variables need to be swapped.
+    if ((step_end - step_start) % 2 == 1) {
+      // this is easily done without stl with memcpy:
+      // memcpy(_device_data.y.data(), y_next.data(), sizeof(real_t) * n_state);
+      real_t *dst = _device_data.y.data();
+      std::copy_n(y_next.data(), n_state * n_particles, dst);
```

Comment:
Still a long copy in the kernel – can we add a note to be careful when porting this to CUDA?
Some discussion of this here: https://stackoverflow.com/a/49037139

We should also look out for this in the profile: how long would a copy of 100,000 particles with 1,000 floats of state take? Could well be longer than the update.
Alternatively just leaving the copy as async may help hide any latency here

Reply:
this is now a swap